segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java

```diff
@@ -84,24 +90,28 @@ private static Map.Entry<Long, Long> applySearch(StreamSegmentStore store, Strin
            return getSegmentLength(store, segment, startIdx);
         }
 
-        return SortUtils.newtonianSearch(idx -> {
-            ReadResult result = store.read(indexSegmentName, idx * NameUtils.INDEX_APPEND_EVENT_SIZE, NameUtils.INDEX_APPEND_EVENT_SIZE, TIMEOUT).join();
-            ReadResultEntry firstElement = result.next();
-            // TODO deal with element which is split over multiple entries.
-            switch (firstElement.getType()) {
-                case Cache: // fallthrough
-                case Storage:
-                    BufferView content = firstElement.getContent().join();
-                    IndexEntry entry = IndexEntry.fromBytes(content);
-                    return entry.getOffset();
-                case Truncated:
-                    throw new SegmentTruncatedException(String.format("Segment %s has been truncated.", segment));
-                case Future:
-                case EndOfStreamSegment:
-                    throw new IllegalStateException(String.format("Unexpected size of index segment of type: %s was encountered for segment %s.", firstElement.getType(), segment));
-            }
-            throw new IllegalStateException(String.format("Unexpected index segment of type: %s was encountered for segment %s.", firstElement.getType(), segment));
-        }, startIdx, endIdx > 0 ? endIdx - 1 : 0, targetOffset, greater);
+        return Retry.withExpBackoff(RETRY_INITIAL_DELAY_MS, RETRY_MULTIPLIER, RETRY_COUNT, RETRY_MAX_DELAY_MS)
+                .retryWhen(ex -> ex instanceof IllegalStateException)
```

Comment:
Can we give this a specific exception. Because in general IllgalStateException should not be retried, and there could be other sources of the exception

Reply:
Added specific exception.